AWS X-Ray support

kf6kjg opened the pull request:
This enables tracking of the entire warmup cycle in X-Ray if it the latter is installed.

kf6kjg commented:
The first commit, prefixed with "SKIP" was me simply working around the fact that the repo currently doesn't follow its own eslint rules. After that, since I'm on my work Windows install, I had to fix the unit test path issues. Finally got to the meat of the PR in the last commit.

And it looks like TravisCI is choking on the Node 8 and 9 test suite, but in areas I've not affected. And passing in all the newer versions as I'd expect it should: I've only changed the lambda, not the plugin.

juanjoDiaz commented on .eslintrc.js:
```diff
@@ -3,4 +3,15 @@ module.exports = {
   env: {
     node: true,
   },
+  rules: {
```
Turning rules off is not OK.

kf6kjg replied:
I completely agree, hence the commit message saying to drop the commit that does this in the merge.

juanjoDiaz commented on test/utils/configUtils.js:
```diff
@@ -1,11 +1,13 @@
+function getServerlessConfig(serverless = {}) {
   return {
-    getProvider: serverless.getProvider || (() => {}),
+    getProvider: serverless.getProvider || (() => { }),
```
Why this space?

kf6kjg replied:
I thought I removed all that auto reformat nonsense before the commit, guess some snuck in anyway. I'll get that out of the PR.

juanjoDiaz commented on test/utils/configUtils.js:
```diff
@@ -1,11 +1,13 @@
     },
     cli: {
-      log() {}
+      log() { }
```
Why this space?

kf6kjg replied:
ditto.

juanjoDiaz commented on src/index.js:
```diff
@@ -152,8 +152,8 @@ class WarmUp {
-      pathFile: `${pathFolder}/index.js`,
-      pathHandler: `${folderName}/index.warmUp`,
+      pathFile: path.join(pathFolder, 'index.js'),
+      pathHandler: path.join(folderName, 'index.warmUp'),
```
`pathHandler` was OK before. Since this is not a path to be used by the OS but a path used by Serverless.

Or am I missing something?

kf6kjg replied:
This commit may have been heavy handed: I was simply bulk changing paths to get the test suite, which are tun by the git hooks, to allow me to push the changes.

juanjoDiaz commented on test/utils/configUtils.js:
```diff
@@ -27,7 +29,7 @@ function getExpectedFunctionConfig(options = {}) {
     description: 'Serverless WarmUp Plugin',
     events: [{ schedule: 'rate(5 minutes)' }],
-    handler: '_warmup/index.warmUp',
+    handler: path.join('_warmup', 'index.warmUp'),
```
As above, this was OK before. Since this is not a path to be used by the OS but a path used by Serverless.

Or am I missing something?

juanjoDiaz commented on test/hook.afterPackageInitialize.test.js:
```diff
@@ -1039,7 +1040,7 @@ describe('Serverless warmup plugin constructor', () => {
     expect(plugin.serverless.service.functions.warmUpPlugin)
       .toEqual(getExpectedFunctionConfig({
-        handler: 'test-folder/index.warmUp',
+        handler: path.join('test-folder', 'index.warmUp'),
```
Same as the other comments. `handler` is always OK with backslashes.

juanjoDiaz commented on src/index.js:
```diff
@@ -311,7 +311,15 @@ class WarmUp {
-const aws = require("aws-sdk");
+let aws;
+try {
+  const AWSXRay = require("aws-xray-sdk-core");
```
How is  `aws-xray-sdk-core` installed in the lambda?

kf6kjg replied:
The same way aws-sdk is: by the user installing it in the project they add this plugin to.

juanjoDiaz replied:
That's not how this plugin works.
`aws-sdk` is always present in the Nodejs lambdas without the user having to install it.

So, as I suspected, this would always throw and use the standard uninstrumented library.

Are you planning to test this in any project of yours?
I have some ideas about how this could work and could create a branch for you to test.
Unfortunately, I don't have time to test this myself at the moment on an actual AWS deployment.

kf6kjg replied:
Yes, I've tested it. :)

juanjoDiaz commented on src/index.js:
```diff
@@ -311,7 +311,15 @@ class WarmUp {
+  aws = AWSXRay.captureAWS(require("aws-sdk"));
+}
+catch {
+  console.log("X-Ray SDK not found, falling back to basic AWS SDK.");
```
I would not notify if X-Ray is not there but the other way around.
I'd notify if X-Ray is used.

kf6kjg replied:
++ That makes sense.

kf6kjg commented:
The first two commits, dealt with in detail below, belong in a different PR IMHO and were only here as a "bandaid" to allow me to commit and push: git's hooks were blocking me due to failing on eslint and the test suite path comparisons.

> The first commit is not OK. It just loosens the rules too much.
Agreed. I'll see if I can get it removed via GitHub's interface. It was added because I didn't feel like doing a massive formatting change in a PR that's supposed to be about just a few lines of added code.

> What kind of errors have you seen in ESLint? I don't see any and the build never broke on Travis.
See [eslintresults_master_b803acb9bf20339caaeb322d689e9048d62a13e9.txt](https://github.com/FidelLimited/serverless-plugin-warmup/files/4123416/eslintresults_master_b803acb9bf20339caaeb322d689e9048d62a13e9.txt)

> The second commit is good. Except for changing the handlers which will break Serverless afaik. I guess that you are using Windows which caused the issue, right?
Yes I'm on Windows at the moment, as I mentioned in the previous comment. This commit, like the first, I'll see if I can drop or you can drop in the merge. See [jesttestresults_master_b803acb9bf20339caaeb322d689e9048d62a13e9.txt](https://github.com/FidelLimited/serverless-plugin-warmup/files/4123495/jesttestresults_master_b803acb9bf20339caaeb322d689e9048d62a13e9.txt) for the problems I was resolving.

> Unfortunately, I don't fully understand how the third commit is supposed to work. How is the X-Ray library installed in the lambda? It seems to me that it will never be there so it will always just use the uninstrumented AWS package.
I simply followed the same thought pattern I found in the code for the AWS SDK: that it's included as part of the user's project and packaged in by serverless. Running under that assumption, and before I submitted this PR, I added this change to the copy I run on one of my APIs and it worked great: I got all the tracking of the calls and fixed a few errors the X-Ray system revealed.

kf6kjg commented:
I'll rebase once #206 is finished and merged.

juanjoDiaz commented:
I've created a branch https://github.com/juanjoDiaz/serverless-plugin-warmup/blob/add_tracing_support/src/index.js of how I would do this in a more reliable way.
Missing documentation and unit tests. Not even manually tested in any way.

To get this in we should update the documentation on how to use tracing and what is needed (tracing option, installed npm module, necessary permissions in the role).

kf6kjg commented:
I see, you're wanting to provide the user with the option of being able to enable or disable it via a config option. I'd chosen not to do that, but it's a good idea: X-Ray can add cost when it is enabled.

I can find no resources that list the SDKs that are preinstalled on AWS Lambdas.  I saw some notes that via layers the user can override the SDK in case the user needs a newer version, but nothing that documented what SDKs, and what version(s), are there by default - perhaps my Google Foo has failed me.

kf6kjg commented:
Is `tracing:bool` is to be applied only in the global section of the config?  I would think that's the case as the code you wrote currently doesn't switch between the two SDKs per lambda call.

kf6kjg commented:
Rebuilt on top of your commit. However I've not yet completed the unit tests and what I did in the last commit didn't feel right: I'm going to have to study the unit test system you have here closer than I have time for ATM.

kf6kjg commented:
Also it looks like Travis is choking on it even though my system is not. Whee.

juanjoDiaz commented:
> Is tracing:bool is to be applied only in the global section of the config? I would think that's the case as the code you wrote currently doesn't switch between the two SDKs per lambda call.

You are right. I shouldn't be hacking after midnight.
I fixed my branch.

> I see, you're wanting to provide the user with the option of being able to enable or disable it via a config option. I'd chosen not to do that, but it's a good idea: X-Ray can add cost when it is enabled.

It's a bit more than that.
It checks whether `aws-xray-sdk-core` is installed and adds it to the lambda package.
If tracing is enabled but `aws-xray-sdk-core` is missing, then there is a warning and the code is not instrumented.

kf6kjg commented:
I don't see any updates to the branches, did you push?

juanjoDiaz commented:
My bad! The push failed because of the push hook.
Not it should be there.
